### PR TITLE
[ONNX] Update embedding_bag export

### PR DESCRIPTION
**Pull request by neginraoof**
Export of embedding bag with dynamic list of offsets.

### CODE REVIEW

File: torch/onnx/symbolic_opset10.py [hunk 1]
```diff
@@ -206,38 +206,12 @@ def embedding_bag(g,
     if scale_grad_by_freq and sym_help._training_mode:
         return sym_help._onnx_unsupported('embedding_bag with scale_grad_by_freq for training mode')
 
-    from torch.onnx.symbolic_opset9 import size, div, select
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset10.py:209]: Why are these removed? Was it that the graph changes certain assumptions don't stand anymore?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -3377,27 +3358,29 @@ def test_embedding_bag(self):
...
 
-    @disableScriptTest()
+    @disableScriptTest()  # scripting prim::Uninitialized
     @skipIfUnsupportedMinOpsetVersion(10)
+    @skipIfUnsupportedOpsetVersion([12])  # Due to ONNX Loop shape inference issue
     def test_embedding_bag_1d_per_sample_weights(self):
         class EmbeddingModel(torch.nn.Module):
             def forward(self, embedding_matrix, input, offset, weights):
-                return torch.nn.functional.embedding_bag(embedding_matrix, input, offsets=offset,
+                return torch.nn.functional.embedding_bag(input, embedding_matrix, offsets=offset,
```
Comment (BowenBao) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3367]: was `torch.nn.functional.embedding_bag` signature updated?

File: torch/onnx/symbolic_opset11.py [hunk 3]
```diff
@@ -732,3 +736,69 @@ def flatten(g, input, start_dim, end_dim):
...
+    offsets_starts = sym_help._slice_helper(g, offsets, axes=[0], starts=[0], ends=[-1], steps=[1])
+    offsets_ends = sym_help._slice_helper(g, offsets, axes=[0], starts=[1], ends=[maxsize], steps=[1])
+
+    loop_len = sym_help._size_helper(g, offsets_ends, g.op("Constant", value_t=torch.tensor(0)))
+    loop = g.op("Loop", loop_len, loop_condition)
+
+    new_block = _add_block(loop.node())
+    block_input_iter = _add_input_to_block(new_block)
+
+    indices_start = _add_node_to_block(new_block, "onnx::Gather", offsets_starts, block_input_iter, axis_i=0)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py:776]: i'm wondering if we can create similar helper function for block, like `g.op` for `Graph` `g`, we'll get `new_block.op('Gather', offsets_starts, ...)`. I think it's ok to omit `onnx` namespace in this case.
Similarly for creating block, and adding input/output to block. But these are less used.

File: torch/onnx/symbolic_opset11.py [hunk 4]
```diff
@@ -732,3 +736,69 @@ def flatten(g, input, start_dim, end_dim):
...
+    # Offsets holds the starting index position of each bag. So we create a list of the indices slices (determined by
+    # offsets) and gather those indices in indices_row. Then we use this subset of indices to gather from embeddings.
+    # The embeddings output is a loop scan output, so we can avoid creating a sequence and inserting elements in.
+    offsets_starts = sym_help._slice_helper(g, offsets, axes=[0], starts=[0], ends=[-1], steps=[1])
+    offsets_ends = sym_help._slice_helper(g, offsets, axes=[0], starts=[1], ends=[maxsize], steps=[1])
+
+    loop_len = sym_help._size_helper(g, offsets_ends, g.op("Constant", value_t=torch.tensor(0)))
+    loop = g.op("Loop", loop_len, loop_condition)
+
+    new_block = _add_block(loop.node())
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py:773]: rename to `loop_block`?

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3367]: Yes. I saw a warning that the previous order is now deprecated.

File: torch/onnx/symbolic_opset10.py (hunk 1 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset10.py:209]: No, actually the previous export logic was buggy for some cases.
The behavior of the embedding_bag functional module is different when input is 2D or 1D. The code here detected this case by checking for this condition in symbolic through accessing the input to indices node, like below:
```if len(indices.node().inputs().__next__().type().sizes()) == 2```
But then I saw other use-cases of embedding_bags, when this condition was a false position.
I'm removing this part of the code (which was mainly for dynamic export of this op with inputs of different shape) since its logic is broken, and I'm adding full dynamic support as part of opset 11.

File: torch/onnx/symbolic_opset11.py (hunk 3 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset11.py]: Thanks. Did this.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -3377,27 +3358,29 @@ def test_embedding_bag(self):
...
         model = EmbeddingModel()
         x = torch.randint(7, (6,))
-        w = torch.randn(6,)
+        w = torch.randn(6, )
         offset = torch.tensor([0, 2, 5])
         embedding_matrix = torch.rand(10, 15)
         self.run_test(model, (embedding_matrix, x, offset, w))
 
-    @disableScriptTest()
+    @disableScriptTest()  # scripting prim::Uninitialized
```
Comment (KsenijaS) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3377]: does prim::Uninitialized appears in the graph for the scripting mode? Does this happen after your changes?

Comment (neginraoof) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3377]: Yes, this is for scripting. I think I encountered it even before this update.

**Review by BowenBao — APPROVED** [review]
LGTM, please rebase to pass unrelated test failures.

**Comment by neginraoof** [comment]
@bzinodev Can we merge this please?
Thanks!

**Review by bzinodev — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -3377,27 +3358,29 @@ def test_embedding_bag(self):
         input = torch.randint(10, (7, 5))
         self.run_test(model, (input))
 
-    @disableScriptTest()
+    @disableScriptTest()  # scripting prim::Uninitialized
```
Comment (KsenijaS) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3361]: nit: prim::Uninitialized is not the only prim op that appears in graph when scripting, maybe add prim:dtype and prim::unchecked_cast in the comment

**Comment by neginraoof** [comment]
cc @bzinodev 
This PR is rebased and ready for merge. Thanks a lot.
